### DISCUSSION TITLE
Installing the integration

### DISCUSSION THREAD

**Discussion by Thailanda** [ekutner/home-connect-hass#141] [Community Support] [answered]
Hello,

I'm trying to install the integration, but I can't get past step 6 in the documentation:

6. "A new window will popup where you will be asked to login to to your Home Connect account and allow Home Assistant to access your appliances. After you approve that you will be redirected back to Home Assistant, continue as instructed."

After I approve I get redirected to my Home Assistant instance with an empty "Error"-Popup.

I have enabled default logging. This is the error message:

`2022-12-15 10:01:28.342 ERROR (MainThread) [aiohttp.server] Error handling request
Traceback (most recent call last):
File "/usr/local/lib/python3.10/site-packages/aiohttp/web_protocol.py", line 435, in _handle_request
resp = await request_handler(request)
File "/usr/local/lib/python3.10/site-packages/aiohttp/web_app.py", line 504, in _handle
resp = await handler(request)
File "/usr/local/lib/python3.10/site-packages/aiohttp/web_middlewares.py", line 117, in impl
return await handler(request)
File "/usr/src/homeassistant/homeassistant/components/http/security_filter.py", line 60, in security_filter_middleware
return await handler(request)
File "/usr/src/homeassistant/homeassistant/components/http/forwarded.py", line 100, in forwarded_middleware
return await handler(request)
File "/usr/src/homeassistant/homeassistant/components/http/request_context.py", line 28, in request_context_middleware
return await handler(request)
File "/usr/src/homeassistant/homeassistant/components/http/ban.py", line 82, in ban_middleware
return await handler(request)
File "/usr/src/homeassistant/homeassistant/components/http/auth.py", line 236, in auth_middleware
return await handler(request)
File "/usr/src/homeassistant/homeassistant/components/http/view.py", line 136, in handle
result = await result
File "/usr/src/homeassistant/homeassistant/components/config/config_entries.py", line 171, in get
return await super().get(request, flow_id)
File "/usr/src/homeassistant/homeassistant/helpers/data_entry_flow.py", line 96, in get
result = await self._flow_mgr.async_configure(flow_id)
File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 280, in async_configure
result = await self._async_handle_step(
File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 367, in _async_handle_step
result: FlowResult = await getattr(flow, method)(user_input)
File "/usr/src/homeassistant/homeassistant/helpers/config_entry_oauth2_flow.py", line 303, in async_step_creation
token = await self.flow_impl.async_resolve_external_data(self.external_data)
File "/usr/src/homeassistant/homeassistant/helpers/config_entry_oauth2_flow.py", line 169, in async_resolve_external_data
return await self._token_request(
File "/usr/src/homeassistant/homeassistant/helpers/config_entry_oauth2_flow.py", line 205, in _token_request
resp.raise_for_status()
File "/usr/local/lib/python3.10/site-packages/aiohttp/client_reqrep.py", line 1004, in raise_for_status
raise ClientResponseError(
aiohttp.client_exceptions.ClientResponseError: 400, message='Bad Request', url=URL('https://api.home-connect.com/security/oauth/token')`

Can someone guide me in the right direction? Am I missing something?

Thanks in advance.

**Comment by ekutner** [comment]
This kind of problem always ends up being a configuration problem. Make sure you carefully follow the instructions

**Reply by Thailanda** [reply]
I did follow the instructions carefully. In my.home-assistant.io i've tried my public url "https://sub.domain.tld" and i also tried the private static ip from my server including port 8123. Both the internal and external URLs open up my home-assistant instance. But I keep getting this error.

I've have tried putting the "client_id" and "client_secret" in the configuration.yaml with double quotes, without quotes and single quotes.

I can control my devices with the home-connect app with wifi disconnected. I've turned the devices (appliances) on and turned wifi on my mobile off. 

Nothings seems to work. Do you need more information to point me to the right direction?

**Comment by ekutner** [comment]
please upload a screenshot of the HC app (hide the secret key) and the final URL you get to after you add the integration, log in to HC and approve the access permissions.

**Comment by Thailanda** [comment]
![HCApp](https://user-images.githubusercontent.com/15957507/207921115-a6f78183-4320-4ffc-9ca6-791bcb2b1c2b.png)
![myHA](https://user-images.githubusercontent.com/15957507/207921133-ab0a20ff-7c64-429f-886b-0f18af10a3ec.png)

After clicking on "LINK ACCOUNT" in the second screenshot I get back to my Home-Assistant instance with an empty Error Popup. The title of the pop-up is just "Error".

I've also tried a different browser. Same issue. :(

**Comment by ekutner** [comment]
This all looks fine.
The instance URL you configure in my.home-assistant.io should be your internal address, with the port number, and it should be the same address you are using to access HA when you are adding the integration.

**Reply by Thailanda** [reply]
I can try that again with the internal address and port. Protocoll would be "http" then and not "https", correct?

**Reply by ekutner** [reply]
yes. but don't change anything in the app, the only change is in the instance URL under my.home-assistant.io

**Reply by Thailanda** [reply]
no difference. Still the same error. :(

Does the client_id and client_secret have to be with or without quotes? If with quotes, double or single? Does it even matter?

**Comment by ekutner** [comment, marked as answer]
the quotes don't matter and if it that was the problem you would have received a different error.
Please upload a screenshot of the HA browser with the URL and the error + a new one the my.home-assistant.io page without erasing the instance URL.

**Reply by ekutner** [reply]
also copy the full URL you get in the my.home-assistant.io because it runs longer than what is visible in the screenshot

**Reply by Thailanda** [reply]
http://192.168.2.174:8123/lovelace/0 thats the requested URL?

![myHAlocal](https://user-images.githubusercontent.com/15957507/207933053-330456df-d996-4448-8808-df043e9f2b40.png)

**Reply by ekutner** [reply]
No, the long URL at the end of the login flow, on the page that has the "link account" button

**Reply by Thailanda** [reply]
https://my.home-assistant.io/redirect/oauth/?code=eyJ4LXJlZyI6IkVVIiwieC1lbnYiOiJQUkQiLCJ0b2tlbiI6IjU1OGY3ZWRhLWVhY2QtNDUyMC1hYTA2LWQyMjdlNGU2MGU2NSIsImNsdHkiOiJwcml2YXRlIn0%3D&state=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJmbG93X2lkIjoiMDU4Y2UwMmQwYjJmMzgwNTMxZGVmYTlmMjdiYjA3ZmUiLCJyZWRpcmVjdF91cmkiOiJodHRwczovL215LmhvbWUtYXNzaXN0YW50LmlvL3JlZGlyZWN0L29hdXRoIn0.eMe6W_bW0WMzsayTuM2VEGBF4DoNF89gfKXnRpIlbQ0&grant_type=authorization_code

**Reply by ekutner** [reply]
this looks fine as well. try completely removing the  integration from HACS, then restart HA then add it back and try again. You configuration seems fine so don't change it.

**Reply by Thailanda** [reply]
Ok, thank you ekutner for your time. Will try that now.

Edit: I've done what you said. Exactly the same behaviour and error. 

Edit2: I've created a second application in home connect. Same error...

**Reply by Thailanda** [reply]
@ekutner any further ideas?

**Reply by ekutner** [reply]
I've looked at this whole thread again and there is one thing I've missed, which is that it fails with a 400 error from Home Connect when Home Assistant is trying to get the access token required to use the API. So I tried testing this with my own account, which worked fine until it stopped working and I started getting the same error you are getting. I suspect that this may be some server protection mechanism gone rogue.

To verify this please add this logging config to your configuration.yaml file:
```
logger:
  default: info
  logs:
    homeassistant.helpers.config_entry_oauth2_flow: debug
```

Then upload the complete log file

**Reply by Thailanda** [reply]
`2022-12-19 22:41:16.681 DEBUG (MainThread) [homeassistant.helpers.config_entry_oauth2_flow] Token request failed with status=400, body={
"error": "invalid_client",
"error_description": "client has limited user list - user not assigned to client"
}
2022-12-19 22:41:16.681 ERROR (MainThread) [aiohttp.server] Error handling request`

I think thats the exciting part of the log. I've googled that error. Suggested fixes were "only small letters in SingleKey ID email", which i fullfill. And the second thing I found was checking the "keep me signed in"-box, which I also tried. Still the same error. 

Do you have any other ideas what could solve my problem?

**Reply by ekutner** [reply]
The error seems to suggest that you don't have a valid "test" user associated with the app. In the "Home Connect user account for testing" option of the app configuration, make sure that you have added the email address of the Home Connect account that you are using in the mobile app, which I assume is also the account where your appliances are registered.

**Reply by Thailanda** [reply]
I already hat that test user set as you can see in my screenshot. I blacked that part in the screenshot. It’s the same email I’m using to control my devices with the mobile app. It’s exactly the same all small letters. My email includes a hyphen (-). Could that be a problem?

**Reply by ekutner** [reply]
I can't say for sure what is the problem but the error is "client has limited user list - user not assigned to client" so you can try playing with that, maybe remove and re-add that user, or try a different user or with the user ID instead of email. Also take a look here: https://github.com/home-assistant/core/issues/83359

I'm just guessing here because the error is related to the configuration of the app and not the integration (as I wrote initially, in the entire history of this integration all these problems have always been on the app). If none of the above works, you can open a ticket with Home Connect support and pray that they respond in a reasonable time.

**Reply by Thailanda** [reply]
Ok, when i click on "change SingleKey ID" the new window shows my email with some capital letters, which caused the problem. I've changed to another email making sure everything is in small letters and now it works. This is really tricky because in Home Connect (Browser [Developer Program] AND mobile App) the email shows with all small letters. Only when clicking on change singlekey id it displays that there are some capital letters in the email for some mysterious reasons.

@ekutner Thank you for your guidance and time. Best regards and have a nice holiday.

**Reply by ekutner** [reply]
Thanks. 
I'm always surprised with the ways this app can fail to work, but happy to see it was resolved.
